keps/sig-auth/3766-referential-authorization/README.md

```diff
@@ -0,0 +1,1212 @@
+        For string
+    }
+    class RoleBinding{
+        Subjects []Subject
+        RoleRefs []RoleRef
+    }
+    class Role{
+        APIGroups []string
+        Resources []string
+        ResourceNames []string
+        Verbs []string
+    }
+```
+
+## Motivation
+
+Any project that wants to enable cross-namespace references currently has to
+choose between introducing a dependency on Gateway API's ReferenceGrant or
+creating a new API that would be partially redundant (leading to confusion for
+users).
+
+Recent interest between SIGs and projects (SIG Network, SIG Storage, and Flux)
+has made it clear that ReferenceGrant is wanted for use cases other than Gateway
+API. We would like to move ReferenceGrant to a neutral home (ideally, under SIG
+Auth) in order to make it the canonical API for managing references across
+namespaces.
+
+Due to lack of better alternatives, Ingress and Gateway controllers, for
+instance, are often deployed with read access to all Secrets in the cluster.
+This can be a significant security risk as these components often already sit on
+the edge of a cluster. Finding a way to compromise an Ingress or Gateway
+controller could potentially provide access to all Secrets in a cluster. This
+KEP provides a more fine-grained approach to authorizing cross- and in-namespace
+references, such that users have the realistic possibility of removing coarse
+RBAC rules such as that the Ingress controller can access all the Secrets in the
+cluster. This KEP does not offer extra security for references, in case RBAC
+rules are applied in such a coarse manner.
+
+### Goals
+
+* Introduce API(s) that enable the same capabilities that ReferenceGrant does
+  within Gateway API.
+* In addition to ReferenceGrant's capabilities, expand the scope of this KEP so
+  that it enables:
+  * Authorizing controllers to read only resources that have been referenced by
+    resources they are implementing.
+  * Building a more generic approach that will enable generic controllers and
+    authorizers to implement this API without prior knowledge of resources that
+    references are from or to.
+* Clearly define how ReferenceGrant should be used, including both current use
+  cases and guidance for future use cases.
+* Implement a library to ensure that ReferenceGrant is implemented consistently
+  by all controllers.
+
+### Non-Goals
+
+* Advanced capabilities like label selectors and the ability to allow references
+  from any namespace are likely reasonable future extensions of this API but are
+  intentionally out of scope for the initial work.
+
+## Proposal
+
+Create 3 new API types in a new `reference.authorization.k8s.io` API Group:
+
+1. **ClusterReferenceGrant:** This resource grants access to resources by
+   following references within the same namespace. For example, a `gateways`
+   ClusterReferenceGrant would grant access to Gateway controllers for
+   same-namespace references from Gateways to TLS Secrets. API developers would
```

Comment:
I still think that we should separate the API object for defining how the API type is shaped (versions, JSONPaths/CELExprs, etc.) and allowing "within namespace references", and not bake them into the same object.

This way one could avoid allowing all within-namespace references, relying only on "individual" ReferenceGrants defined in the cluster (for specific within and cross-namespace references).